### PR TITLE
ENYO-2205 Initialize selected items when collection changed

### PR DESCRIPTION
**Pull request by webOS101**
* Issue *
Selection handling in DataRepeater/DataList was inconsistent. Selections would not be properly tracked with DataList and selections were not reset when changing collections.

* Solution *
If a `selectionProperty` is defined, filter out selected items to initialize the internal `_selection` property.

Enyo-DCO-1.1-Signed-off-by: Roy Sutton <roy.sutton@lge.com>

### CODE REVIEW

File: lib/DataRepeater.js [hunk 1]
```diff
@@ -454,7 +454,7 @@ var DataRepeater = module.exports = kind(
 	* @private
 	*/
 	initCollection: function (c, p) {
-		var e;
+		var e, isArry = false;
```
Comment (aarontam) [inline lib/DataRepeater.js]: Can we make this `isArray` or `isArr`? :smile:

**Comment by webOS101** [comment]
Updated variable name and discovered test is not very good so updated the test.

File: lib/DataRepeater.js [hunk 2]
```diff
@@ -471,6 +473,14 @@ var DataRepeater = module.exports = kind(
 				p.removeListener(e, this._handlers[e]);
 			}
 		}
+		if(c && this.selectionProperty) {
+			var filter = isArray ?
+				function(record) { return record[this.selectionProperty]; } :
+				function(record) { return record.get(this.selectionProperty); };
```
Comment (rajyavardhanp) [inline lib/DataRepeater.js]: functions should be bound to the scope to identify the scope's properties inside I guess.

File: lib/DataRepeater.js [hunk 3]
```diff
@@ -463,6 +463,8 @@ var DataRepeater = module.exports = kind(
 		// Decorate native JS array with at() so that we can
 		// access members of our dataset consistently, regardless
 		// of whether our data is in an array or a Collection
+		// TODO: Document that we add a method to an array or use
```
Comment (ryanjduffy) [inline lib/DataRepeater.js]: Rather than documenting this, let's just fix it.

`Object.defineProperty(c, 'at', {value: at, enumerable: false})`

File: lib/DataRepeater.js [hunk 4]
```diff
@@ -215,9 +215,9 @@ var DataRepeater = module.exports = kind(
 	create: kind.inherit(function (sup) {
 		return function () {
 			sup.apply(this, arguments);
+			this.selectionType = this.multipleSelection ? (this.groupSelection ? 'group' : 'multi') : this.selectionType;
 			this.collectionChanged();
 			// Converting deprecated selection properties to our current selection API
-			this.selectionType = this.multipleSelection ? (this.groupSelection ? 'group' : 'multi') : this.selectionType;
```
Comment (ryanjduffy) [inline lib/DataRepeater.js]: Why was this moved up?

File: lib/DataRepeater.js [hunk 5]
```diff
@@ -471,6 +473,17 @@ var DataRepeater = module.exports = kind(
 				p.removeListener(e, this._handlers[e]);
 			}
 		}
+		if(c && this.selectionProperty) {
```
Comment (ryanjduffy) [inline lib/DataRepeater.js]: Space between `if` and `(`

File: lib/DataRepeater.js [hunk 6]
```diff
@@ -471,6 +473,17 @@ var DataRepeater = module.exports = kind(
...
+			var selectionProperty = this.selectionProperty,
```
Comment (ryanjduffy) [inline lib/DataRepeater.js]: While we're at it, let's move these `var`s up to the top of the method. Might consider defining these methods at the top of the module (like `at()`).

File: lib/DataRepeater.js (hunk 4 above)
Comment (webOS101) [inline lib/DataRepeater.js]: Because selectionType needs to be set before the collectionChanged method is invoked.

File: lib/DataRepeater.js (hunk 6 above)
Comment (webOS101) [inline lib/DataRepeater.js]: Good idea.  I have done so, and moved (and renamed) the local shadow of selectionProperty

File: lib/DataRepeater.js (hunk 4 above)
Comment (aarontam) [inline lib/DataRepeater.js]: Should we move the accompanying comment, as well, then?

Comment (webOS101) [inline lib/DataRepeater.js]: (rageface)  yes

Comment (aarontam) [inline lib/DataRepeater.js]: Actually, does not seem like we need to move this line - going to move it back and save the comment from having to pack its things.

File: lib/DataRepeater.js [hunk 7]
```diff
@@ -464,13 +476,20 @@ var DataRepeater = module.exports = kind(
...
-			c.at = at.bind(c);
+			Object.defineProperty(c, 'at', {value: at, enumerable: false});
 		}
 		if (p && p.removeListener) {
 			for (e in this._handlers) {
 				p.removeListener(e, this._handlers[e]);
 			}
 		}
+		if (c && this.selectionProperty) {
+			filterSelectionProperty = this.selectionProperty;
```
Comment (aarontam) [inline lib/DataRepeater.js]: Going to make a small tweak so that we don't need to maintain this var, and instead will pass the context to the `filter` method.
